File: UnitTests/LUKeychainAccessSpec.m
#import <LUKeychainAccess/LUKeychainAccess.h>
#import <Kiwi/Kiwi.h>
#import "LUTestErrorHandler.h"
#import "LUTestNonNSCodingCompliantObject.h"
#import "LUTestNSCodingCompliantObject.h"
#import "NSKeyedArchiver+Additions.h"

SPEC_BEGIN(LUKeychainAccessSpec)

describe(@"LUKeychainAccess", ^{
  __block LUKeychainAccess *keychainAccess;
  __block LUKeychainServices *keychainServices;
  __block LUTestErrorHandler *errorHandler;
  NSString *testGroup = @"test_group";
  NSDictionary *testAdditionalParams = @{ @"test":@"test" };

  // Helpers
  id (^errorReturningBlock)(NSArray *) = ^id(NSArray *params) {
    __autoreleasing NSError **error;
    [(NSValue *)[params lastObject] getValue:&error];
    *error = [NSError errorWithDomain:LUKeychainAccessErrorDomain code:errSecDuplicateItem userInfo:nil];

    return [KWValue valueWithBool:NO];
  };

  beforeEach(^{
    errorHandler = [[LUTestErrorHandler alloc] init];

    keychainServices = [LUKeychainServices nullMock];
    [LUKeychainServices stub:@selector(keychainServices) andReturn:keychainServices];

    keychainAccess = [LUKeychainAccess standardKeychainAccess];
    keychainAccess.errorHandler = errorHandler;
  });

  // Public Methods

  describe(@"standardKeychainAccess", ^{
    it(@"returns a LUKeychainAccess", ^{
      [[keychainAccess should] beKindOfClass:[LUKeychainAccess class]];
    });

    it(@"returns a new LUKeychainAccess each time", ^{
      LUKeychainAccess *keychainAccess1 = [LUKeychainAccess standardKeychainAccess];
      LUKeychainAccess *keychainAccess2 = [LUKeychainAccess standardKeychainAccess];

      [[keychainAccess1 shouldNot] equal:keychainAccess2];
    });
  });

  describe(@"deleteAll", ^{
    it(@"deletes all items from the keychain", ^{
      [[keychainServices should] receive:@selector(deleteAllItemsWithError:)];

      [keychainAccess deleteAll];
    });

    context(@"if the delete fails", ^{
      beforeEach(^{
        [keychainServices stub:@selector(deleteAllItemsWithError:) withBlock:errorReturningBlock];
      });

      it(@"notifies the error handler", ^{
        [keychainAccess deleteAll];

        [[errorHandler.lastError shouldNot] beNil];
      });
    });
  });

  describe(@"deleteObjectForKey:", ^{
    NSString *testKeyToDelete = @"test_key_to_delete";

    it(@"deletes the item with the given key from the keychain", ^{
      [[keychainServices should] receive:@selector(deleteItemWithKey:error:)
                           withArguments:testKeyToDelete, KWAny.any];

      [keychainAccess deleteObjectForKey:testKeyToDelete];
    });

    context(@"if the delete fails", ^{
      beforeEach(^{
        [keychainServices stub:@selector(deleteItemWithKey:error:) withBlock:errorReturningBlock];
      });

      it(@"notifies the error handler", ^{
        [keychainAccess deleteObjectForKey:testKeyToDelete];
      });
    });
  });

  // Getters

  describe(@"accessGroup", ^{
    beforeEach(^{
      [keychainServices stub:@selector(accessGroup) andReturn:testGroup];
    });

    it(@"returns the accessGroup of keychain services", ^{
      [[keychainAccess.accessGroup should] equal:testGroup];
    });
  });

  describe(@"additionalQueryParams", ^{
    beforeEach(^{
      [keychainServices stub:@selector(additionalQueryParams) andReturn:testAdditionalParams];
    });

    it(@"returns the additionalQueryParams of keychain services", ^{
      [[keychainAccess.additionalQueryParams should] equal:testAdditionalParams];
    });
  });

  describe(@"boolForKey:", ^{
    NSString *key = @"boolTest";

    it(@"returns the boolValue of the object stored at the key", ^{
      BOOL testBool = YES;
      [keychainAccess stub:@selector(objectForKey:ofClass:) andReturn:@(testBool) withArguments:key, NSNumber.class];

      [[theValue([keychainAccess boolForKey:key]) should] equal:theValue(testBool)];
    });
  });

  describe(@"dataForKey:", ^{
    NSData *expectedResult = [@"value" dataUsingEncoding:NSUTF8StringEncoding];
    NSString *key = @"dataTest";

    it(@"returns the data stored in keychain at the key", ^{
      [[keychainServices should] receive:@selector(dataForKey:error:)
                               andReturn:expectedResult
                           withArguments:key, KWAny.any];

      [[[keychainAccess dataForKey:key] should] equal:expectedResult];
    });

    context(@"if the services command fails", ^{
      beforeEach(^{
        [keychainServices stub:@selector(dataForKey:error:) withBlock:^id(NSArray *params) {
          errorReturningBlock(params);
          return nil;
        }];
      });

      it(@"notifies the error handler", ^{
        [keychainAccess dataForKey:key];

        [[errorHandler.lastError shouldNot] beNil];
      });
    });
  });

  describe(@"doubleForKey:", ^{
    NSString *key = @"doubleTest";

    it(@"returns the doubleValue of the object stored at the key", ^{
      double testDouble = 123.0;
      [keychainAccess stub:@selector(objectForKey:ofClass:) andReturn:@(testDouble) withArguments:key, NSNumber.class];

      [[theValue([keychainAccess doubleForKey:key]) should] equal:theValue(testDouble)];
    });
  });

  describe(@"floatForKey:", ^{
    NSString *key = @"floatTest";

    it(@"returns the floatValue of the object stored at the key", ^{
      float testFloat = 123.0;
      [keychainAccess stub:@selector(objectForKey:ofClass:) andReturn:@(testFloat) withArguments:key, NSNumber.class];

      [[theValue([keychainAccess floatForKey:key]) should] equal:theValue(testFloat)];
    });
  });

  describe(@"integerForKey:", ^{
    NSString *key = @"integerTest";

    it(@"returns the integerValue of the object stored at the key", ^{
      NSInteger testInteger = 123;
      [keychainAccess stub:@selector(objectForKey:ofClass:) andReturn:@(testInteger) withArguments:key, NSNumber.class];

      [[theValue([keychainAccess integerForKey:key]) should] equal:theValue(testInteger)];
    });
  });

  describe(@"stringForKey:", ^{
    NSString *key = @"stringTest";

    it(@"returns a UTF-8 encoded string from the data stored at the key", ^{
      NSString *testString = @"testString";
      [keychainAccess stub:@selector(dataForKey:)
                 andReturn:[testString dataUsingEncoding:NSUTF8StringEncoding]
             withArguments:key];

      [[[keychainAccess stringForKey:key] should] equal:testString];
    });
  });

  describe(@"objectForKey:ofClass:", ^{
    NSString *key = @"objectTest";

    it(@"returns the unarchived object from the data stored at the key", ^{
      NSArray *testObject = @[@1, @2];
      [keychainAccess stub:@selector(dataForKey:)
                 andReturn:[NSKeyedArchiver lu_archivedDataWithRootObject:testObject]
             withArguments:key];

      [[[keychainAccess objectForKey:key ofClass:[NSArray class]] should] equal:testObject];
    });

    context(@"when the object is not NSCoding compliant", ^{
      it(@"throws an exception", ^{
        LUTestNonNSCodingCompliantObject *testObject = [[LUTestNonNSCodingCompliantObject alloc] init];
        testObject.testProperty1 = 2;
        testObject.testProperty2 = @"Test";

        [[theBlock(^{
          [keychainAccess setObject:testObject forKey:key];
        }) should] raiseWithName:@"Archiver Error"];
      });
    });

    context(@"when the object is not NSSecureCoding compliant", ^{
      it(@"continues to store the archived data of the object with the key", ^{
        LUTestNSCodingCompliantObject *testObject = [[LUTestNSCodingCompliantObject alloc] init];
        testObject.testProperty1 = 2;
        testObject.testProperty2 = @"Test";

        [[keychainAccess should] receive:@selector(setData:forKey:)
                           withArguments:[NSKeyedArchiver lu_archivedDataWithRootObject:testObject], key];

        [keychainAccess setObject:testObject forKey:key];
      });
    });
  });

  describe(@"recursivelyFindObjectForKey:fromClass:", ^{
    NSString *key = @"objectTest";
    NSArray *testObject = @[@1, @2];

    context(@"when the object has been added", ^{
      it(@"it finds the object recursively", ^{
        [keychainAccess stub:@selector(dataForKey:)
                   andReturn:[NSKeyedArchiver lu_archivedDataWithRootObject:testObject]
               withArguments:key];

        [[[keychainAccess objectForKey:key ofClass:[testObject class]] should] beNil];
        [[[keychainAccess recursivelyFindObjectForKey:key fromClass:[testObject class]] should] equal:testObject];
      });
    });

    context(@"when the object has not been added", ^{
      it(@"returns nil", ^{
        [[[keychainAccess objectForKey:key ofClass:[testObject class]] should] beNil];
        [[[keychainAccess recursivelyFindObjectForKey:key fromClass:[testObject class]] should] beNil];
      });
    });
  });

  // Setters

  describe(@"registerDefaults:", ^{
    beforeEach(^{
      [keychainAccess stub:@selector(objectForKey:ofClass:)];
      [keychainAccess stub:@selector(setString:forKey:)];
      [keychainAccess stub:@selector(setObject:forKey:)];
    });

    it(@"sets strings using setString: instead of setObject:", ^{
      NSString *newValue = @"newValue";
      [[keychainAccess should] receive:@selector(setString:forKey:) withArguments:newValue, @"newKey"];
      [keychainAccess registerDefaults:@{@"newKey": newValue}];
    });

    it(@"sets other data types via setObject:", ^{
      [[keychainAccess should] receive:@selector(setObject:forKey:) withArguments:@1, @"newKey"];
      [keychainAccess registerDefaults:@{@"newKey" : @1}];
    });

    it(@"doesn't overwrite existing values", ^{
      [keychainAccess clearStubs];
      [keychainAccess stub:@selector(objectForKey:ofClass:) andReturn:@"existingValue" withArguments:@"existingKey", NSNumber.class];

      [[keychainAccess shouldNot] receive:@selector(setObject:forKey:) withArguments:@YES, @"existingKey"];
      [keychainAccess registerDefaults:@{@"existingKey" : @YES}];

      [keychainAccess clearStubs];
      [keychainAccess stub:@selector(stringForKey:) andReturn:@"existingValue" withArguments:@"existingKey"];
      [[keychainAccess shouldNot] receive:@selector(setString:forKey:) withArguments:@"newValue", @"existingKey"];

      [keychainAccess registerDefaults:@{@"existingKey" : @"newValue"}];
    });

    it(@"is able to check existing values recursively", ^{
      [LUKeychainServices stub:@selector(keychainServices) andReturn:[LUKeychainServices new]];
      keychainAccess = [LUKeychainAccess standardKeychainAccess];

      [keychainAccess setObject:@NO forKey:@"existingBool"];
      [keychainAccess setObject:@[@"existingValue"] forKey:@"existingArray"];
      [keychainAccess setObject:[NSSet setWithArray:@[@"existingValue"]] forKey:@"existingSet"];
      [keychainAccess setObject:@{@"key":@"existingValue"} forKey:@"existingDic"];
      [keychainAccess setString:@"existingValue" forKey:@"existingString"];

      NSDictionary *dictionary = @{@"existingBool": @YES,
                                   @"newBool": @YES,
                                   @"existingArray": @[@"newValue"],
                                   @"newArray": @[@"newValue"],
                                   @"existingSet": [NSSet setWithArray:@[@"newValue"]],
                                   @"newSet": [NSSet setWithArray:@[@"newValue"]],
                                   @"existingDic": @{@"key": @"newValue"},
                                   @"newDic": @{@"key": @"newValue"},
                                   @"existingString":@"newValue",
                                   @"newString":@"newValue"
      };

      [keychainAccess registerDefaults:dictionary];

      [[theValue([keychainAccess boolForKey:@"existingBool"]) should] beNo];
      [[theValue([keychainAccess boolForKey:@"newBool"]) should] beYes];
      [[[keychainAccess objectForKey:@"existingArray" ofClass:[NSArray class]] should] equal:@[@"existingValue"]];
      [[[keychainAccess objectForKey:@"newArray" ofClass:[NSArray class]] should] equal:@[@"newValue"]];
      [[[keychainAccess objectForKey:@"existingSet" ofClass:[NSSet class]] should]
       equal:[NSSet setWithArray:@[@"existingValue"]]];
      [[[keychainAccess objectForKey:@"newSet" ofClass:[NSSet class]] should]
       equal:[NSSet setWithArray:@[@"newValue"]]];
      [[[keychainAccess objectForKey:@"existingDic" ofClass:[NSDictionary class]] should]
       equal:@{@"key":@"existingValue"}];
      [[[keychainAccess objectForKey:@"newDic" ofClass:[NSDictionary class]] should] equal:@{@"key":@"newValue"}];
      [[[keychainAccess stringForKey:@"existingString"] should] equal:@"existingValue"];
      [[[keychainAccess stringForKey:@"newString"] should] equal:@"newValue"];
    });

    it(@"sets the value for new keys", ^{
      [[keychainAccess should] receive:@selector(setObject:forKey:) withArguments:@YES, @"newKey"];
      [keychainAccess registerDefaults:@{@"newKey" : @YES}];
    });

    it(@"can set multiple keys", ^{
      [[keychainAccess should] receive:@selector(setObject:forKey:) withArguments:@YES, @"foo"];
      [[keychainAccess should] receive:@selector(setObject:forKey:) withArguments:@100, @"bar"];
      [keychainAccess registerDefaults:@{@"foo" : @YES, @"bar" : @100}];
    });
  });

  describe(@"setAccessGroup:", ^{
    it(@"sets the access group on the keychain services", ^{
      [[keychainServices should] receive:@selector(setAccessGroup:) withArguments:testGroup];
      keychainAccess.accessGroup = testGroup;
    });
  });

  describe(@"setAdditionalQueryParams:", ^{
    it(@"sets the additionalQueryParams dictionary on the keychain services", ^{
      [[keychainServices should] receive:@selector(setAdditionalQueryParams:) withArguments:testAdditionalParams];
      keychainAccess.additionalQueryParams = testAdditionalParams;
    });
  });

  describe(@"setBool:forKey:", ^{
    NSString *key = @"boolTest";

    it(@"stores the object version of the value with the key", ^{
      BOOL testBool = NO;
      [[keychainAccess should] receive:@selector(setObject:forKey:) withArguments:@(testBool), key];

      [keychainAccess setBool:testBool forKey:key];
    });
  });

  describe(@"setData:forKey:", ^{
    NSString *key = @"dataTest";

    context(@"when the data is nil", ^{
      it(@"deletes the data at the key", ^{
        [[keychainServices should] receive:@selector(deleteItemWithKey:error:) withArguments:key, KWAny.any];

        [keychainAccess setData:nil forKey:key];
      });
    });

    context(@"when the data is non-nil", ^{
      NSData *testData = [@"testData" dataUsingEncoding:NSUTF8StringEncoding];

      it(@"attempts to add the data to the Keychain", ^{
        [[keychainServices should] receive:@selector(addData:forKey:error:) withArguments:testData, key, KWAny.any];

        [keychainAccess setData:testData forKey:key];
      });

      context(@"when the item already exists in the keychain", ^{
        beforeEach(^{
          [keychainServices stub:@selector(addData:forKey:error:) withBlock:errorReturningBlock];
        });

        it(@"updates the item with the new value", ^{
          [[keychainServices should] receive:@selector(updateData:forKey:error:)
                               withArguments:testData, key, KWAny.any];

          [keychainAccess setData:testData forKey:key];
        });

        context(@"if the services command fails", ^{
          beforeEach(^{
            [keychainServices stub:@selector(updateData:forKey:error:) withBlock:errorReturningBlock];
          });

          it(@"notifies the error handler", ^{
            [keychainAccess setData:testData forKey:key];

            [[errorHandler.lastError shouldNot] beNil];
          });
        });
      });
    });
  });

  describe(@"setDouble:forKey:", ^{
    NSString *key = @"doubleTest";

    it(@"stores the object version of the value with the key", ^{
      double testDouble = 123.0f;
      [[keychainAccess should] receive:@selector(setObject:forKey:) withArguments:@(testDouble), key];

      [keychainAccess setDouble:testDouble forKey:key];
    });
  });

  describe(@"setFloat:forKey:", ^{
    NSString *key = @"floatTest";

    it(@"stores the object version of the value with the key", ^{
      float testFloat = 123.0;
      [[keychainAccess should] receive:@selector(setObject:forKey:) withArguments:@(testFloat), key];

      [keychainAccess setFloat:testFloat forKey:key];
    });
  });

  describe(@"setInteger:forKey:", ^{
    NSString *key = @"integerTest";

    it(@"stores the object version of the value with the key", ^{
      NSInteger testInteger = 123;
      [[keychainAccess should] receive:@selector(setObject:forKey:) withArguments:@(testInteger), key];

      [keychainAccess setInteger:testInteger forKey:key];
    });
  });

  describe(@"setString:forKey:", ^{
    NSString *key = @"stringTest";

    it(@"stores the string encoded as UTF-8 with the key", ^{
      NSString *testString = @"testString";
      [[keychainAccess should] receive:@selector(setData:forKey:)
                         withArguments:[testString dataUsingEncoding:NSUTF8StringEncoding], key];

      [keychainAccess setString:testString forKey:key];
    });
  });

  describe(@"setObject:forKey:", ^{
    NSString *key = @"objectTest";

    it(@"stores the archived data of the object with the key", ^{
      NSArray *testObject = @[@1, @2];
      [[keychainAccess should] receive:@selector(setData:forKey:)
                         withArguments:[NSKeyedArchiver lu_archivedDataWithRootObject:testObject], key];

      [keychainAccess setObject:testObject forKey:key];
    });

    context(@"when the object is not NSCoding compliant", ^{
      it(@"throws an exception", ^{
        LUTestNonNSCodingCompliantObject *testObject = [[LUTestNonNSCodingCompliantObject alloc] init];
        testObject.testProperty1 = 2;
        testObject.testProperty2 = @"Test";

        [[theBlock(^{
          [keychainAccess setObject:testObject forKey:key];
        }) should] raiseWithName:@"Archiver Error"];
      });
    });

    context(@"when the object is not NSSecureCoding compliant", ^{
      it(@"continues to store the archived data of the object with the key", ^{
        LUTestNSCodingCompliantObject *testObject = [[LUTestNSCodingCompliantObject alloc] init];
        testObject.testProperty1 = 2;
        testObject.testProperty2 = @"Test";

        [[keychainAccess should] receive:@selector(setData:forKey:)
                           withArguments:[NSKeyedArchiver lu_archivedDataWithRootObject:testObject], key];

        [keychainAccess setObject:testObject forKey:key];
      });
    });
  });
});

SPEC_END

File: Swift UnitTests/LUKeychainAccessSpec.swift
//
//  LUKeychainAccessSpec.swift
//  Swift UnitTests
//
//  Copyright © 2022 GrubHub. All rights reserved.
//

import XCTest
import Foundation
import Nimble
import Quick
import ObjectiveC
@testable import LUKeychainAccess

struct Observer {
  struct Key {
    static let Add = "add"
    static let GetData = "getData"
    static let DeleteAll = "deleteAll"
    static let DeleteItem = "deleteItem"
    static let Update = "update"
    static let Data = "data"
    static let Key = "key"
  }
  
  var record: [String: Dictionary<String, Any>] = [:]
  
  func functionWasCalled(key: String) -> Bool {
    return record[key] != nil
  }
}

final class LUKeychainAccessSpec: QuickSpec {
  var servicesObserver: Observer = Observer()
  
  override func spec() {
    var access: LUKeychainAccess!
    var mockServices: KeychainServicesMock!
    var errorHandler: ErrorHandler!
    let testGroup = "test_group"
    let testAdditionalParams = ["test": "test"]
    let testError = NSError(domain: "test", code: 32)
    
    beforeEach {
      errorHandler = ErrorHandler()
      access = LUKeychainAccess.standardKeychainAccess
      access.errorHandler = errorHandler
      mockServices = KeychainServicesMock()
      access.services = mockServices
      mockServices.observer = self
      self.servicesObserver.record = [:]
    }
    
    // MARK: - standardKeychainAccess
    describe("standardKeychainAccess") {
      it("returns a LUKeychainAccess") {
        expect(access).to(beAKindOf(LUKeychainAccess.self))
      }
      
      it("returns a new LUKeychainAccess each time") {
        let access1 = LUKeychainAccess.standardKeychainAccess
        let access2 = LUKeychainAccess.standardKeychainAccess
        
        expect(access1).toNot(equal(access2))
      }
    }
    
    // MARK: - Deletes
    describe("deleteAll") {
      it("deleteAll") {
        access.deleteAll()
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.DeleteAll)).to(beTrue())
      }
      
      context("if the delete fails") {
        beforeEach {
          mockServices.stubDeleteAll(with: testError, result: false)
        }
        
        it("notifies the error handler") {
          access.deleteAll()
          expect(errorHandler.lastError).toNot(beNil())
          expect((errorHandler.lastError! as NSError).code).to(equal(testError.code))
        }
      }
    }
    
    describe("deleteObjectForkey:") {
      let testKeyToDelete = "test_key_to_delete"
      it("deletes the item with the given key from the keychain") {
        access.deleteObject(for: testKeyToDelete)
        
        if let arguments = self.servicesObserver.record[Observer.Key.DeleteItem],
           let keyArg: String = arguments[Observer.Key.Key] as? String {
          expect(keyArg).to(equal(testKeyToDelete))
        } else {
          fail("deleteObject was not called with key")
        }
      }
      
      context("if the delete fails") {
        beforeEach {
          mockServices.stubDeleteItem(with: testError, result: false)
        }
        
        it("notifies the error handler") {
          access.deleteObject(for: testKeyToDelete)
          expect(errorHandler.lastError).toNot(beNil())
          expect((errorHandler.lastError! as NSError).code).to(equal(testError.code))
        }
      }
    }
    
    // MARK: - Getters
    describe("accessGroup") {
      beforeEach {
        mockServices.stubbedAccessGroup = testGroup
      }
      
      it("returns the accessGroup of keychain services") {
        expect(access.accessGroup).to(equal(testGroup))
      }
    }
    
    describe("additionalQueryParams") {
      beforeEach {
        mockServices.stubbedAdditionalQueryParams = testAdditionalParams
      }
      
      it("returns the additionalQueryParams of keychain services") {
        expect(access.additionalQueryParams as? [String: String]).to(equal(testAdditionalParams))
      }
    }
    
    describe("boolForKey:") {
      let key = "boolTest"
      let testBool = true
      let testBoolData = data(for: NSNumber(value: testBool))
      
      it("returns the boolValue of the object stored at the key") {
        mockServices.stubData(with: nil, result: testBoolData)
        expect(access.bool(for: key)).to(equal(testBool))
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.GetData)).to(beTrue())
      }
    }
    
    describe("dataForKey:") {
      let expectedResult = "value".data(using: String.Encoding.utf8)!
      let key = "dataTest"
      
      it("returns the data stored in keychain at the key") {
        mockServices.stubData(with: nil, result: expectedResult)
        expect(access.data(for: key)).to(equal(expectedResult))
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.GetData)).to(beTrue())
      }
      
      context("if the services command fails") {
        beforeEach {
          mockServices.stubData(with: testError, result: nil)
        }
        
        it("notifies the error handler") {
          expect(access.data(for: key)).to(beNil())
          expect(errorHandler.lastError).toNot(beNil())
          expect((errorHandler.lastError! as NSError).code).to(equal(testError.code))
        }
      }
    }
    
    describe("doubleForKey:") {
      let key = "doubleTest"
      let testValue: Double = 123.0
      let testData: Data? = data(for: NSNumber(value: testValue))
      
      it("returns the doubleValue of the object stored at the key") {
        mockServices.stubData(with: nil, result: testData)
        expect(access.double(for: key)).to(equal(testValue))
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.GetData)).to(beTrue())
      }
    }
    
    describe("floatForKey:") {
      let key = "floatTest"
      let testValue: Float = 123.0
      let testData: Data? = data(for: NSNumber(value: testValue))
      
      it("returns the floatValue of the object stored at the key") {
        mockServices.stubData(with: nil, result: testData)
        expect(access.float(for: key)).to(equal(testValue))
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.GetData)).to(beTrue())
      }
    }
    
    describe("integerForKey:") {
      let key = "integerTest"
      let testValue: Int = 123
      let testData: Data? = data(for: NSNumber(value: testValue))
      
      it("returns the integerValue of the object stored at the key") {
        mockServices.stubData(with: nil, result: testData)
        expect(access.integer(for: key)).to(equal(testValue))
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.GetData)).to(beTrue())
      }
    }
    
    describe("stringForKey:") {
      let key = "stringTest"
      let testValue = "testString"
      let testData = testValue.data(using: String.Encoding.utf8)
      
      it("returns a UTF-8 encoded string form teh data stored at the key") {
        mockServices.stubData(with: nil, result: testData)
        expect(access.string(for: key)).to(equal(testValue))
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.GetData)).to(beTrue())
      }
    }
    
    describe("objectForKey:ofClasses") {
      let key = "objectTest"
      let testValue = [NSNumber(value: 1), NSNumber(value: 2)]
      let testData = NSKeyedArchiver.lu_archivedData(with: testValue)
      
      it("returns the unarchived object from the data stored at the key") {
        mockServices.stubData(with: nil, result: testData)
        expect(access.object(for: key, ofClasses: [NSArray.self]) as? [NSNumber]).to(equal(testValue))
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.GetData)).to(beTrue())
      }
      
      context("when the incorrect class types are passed in") {
        let set = Set(["A", "B", "C"])
        let testObject = [set]
        let testData = NSKeyedArchiver.lu_archivedData(with: testObject)
        
        it("notifies the error handler") {
          mockServices.stubData(with: nil, result: testData)
          expect(access.object(for: key, ofClasses: [NSArray.self])).to(beNil())
          expect(errorHandler.lastError).toNot(beNil())
        }
      }
    }
    
    describe("objectForKey:ofClass") {
      let key = "objectTest"
      let testValue = [NSNumber(value: 1), NSNumber(value: 2)]
      let testData = NSKeyedArchiver.lu_archivedData(with: testValue)
      
      it("returns the unarchived object from the data stored at the key") {
        mockServices.stubData(with: nil, result: testData)
        expect(access.object(for: key, ofClass: NSArray.self) as? [NSNumber]).to(equal(testValue))
        expect(self.servicesObserver.functionWasCalled(key: Observer.Key.GetData)).to(beTrue())
      }
      
      context("when the object is not NSCoding compliant") {
        let testObject = NSCodingNonCompliantObject()
        testObject.testProperty2 = "test"
        testObject.testProperty1 = 2
        
        it("notify the error handler") {
          expect(access.set(object: testObject, for: key)).to(beFalse())
          expect(errorHandler.lastError).toNot(beNil())
        }
      }
      
//      context("when the object is not NSSecureCoding compliant") {
//        let testObject = NSCodingCompliantObject(coder: NSCoder())
//        testObject!.testProperty2 = "test"
//        testObject!.testProperty1 = 2
//        let testData = data(for: testObject!)!
//
//        it("continues to store the archived data of the object with the key") {
//          mockServices.stubUpdate(with: nil, result: testData)
//          expect(access.set(object: testObject!, for: key)).to(beTrue())
//          expect(errorHandler.lastError).to(beNil())
//        }
//      }
    }
    
    describe("recursivelyFindObjectForKey:fromClass:") {
      let key = "objectTest"
      let testValue = [NSNumber(value: 1), NSNumber(value: 2)]
      let testObject = data(for: testValue)!
      
      context("when the object has been added") {
        it("finds the object recursively") {
          mockServices.stubData(with: nil, result: testObject)
          expect(access.object(for: key, ofClass: NSArray.self) as? [NSNumber]).to(equal(testValue))
          expect(access.recursivelyFindObject(for: key, from: NSArray.self) as? [NSNumber]).to(equal(testValue))
        }
      }
      
      context("when the object has not been added") {
        it("returns nil") {
          expect(access.object(for: key, ofClass: NSArray.self)).to(beNil())
          expect(access.recursivelyFindObject(for: key, from: NSArray.self)).to(beNil())
        }
      }
    }
    
    // MARK: - Setters
    describe("registerDefaults:") {
      beforeEach {
        access.services = LUKeychainServices()
      }
      
      it("doesn't overwrite existing values") {
        
        access.register(defaults: ["key": "value1"])
        access.register(defaults: ["key": "value2"])
        expect(access.string(for: "key")).to(equal("value1"))
      }
      
      it("is able to check existing values recursively") {
        access.services = LUKeychainServices()
        access.set(bool: false, for: "existingBool")
        access.set(object: ["existingValue"], for: "existingArray")
        let set: Set = ["existingValue"]
        access.set(object: set, for: "existingSet")
        access.set(object: ["key":"existingValue"], for: "existingDic")
        access.set(string: "existingValue", for: "existingString")
        
        let newSet: Set = ["newValue"]
        let dictionary: Dictionary<String, Any> =
        ["existingBool": true,
         "newBool": true,
         "existingArray": ["newValue"],
         "newArray": ["newValue"],
         "existingSet": newSet,
         "newSet": newSet,
         "existingDic": ["key":"newValue"],
         "newDic": ["key":"newValue"],
         "existingString": "newValue",
         "newString": "newValue"]
        
        access.register(defaults: dictionary)
        
        expect(access.bool(for: "existingBool")).to(beFalse())
        expect(access.bool(for: "newBool")).to(beTrue())
        expect(access.object(for: "existingArray", ofClass: NSArray.self) as? [String]).to(equal(["existingValue"]))
        expect(access.object(for: "newArray", ofClass: NSArray.self) as? [String]).to(equal(["newValue"]))
        expect(access.object(for: "existingSet", ofClass: NSSet.self) as? Set).to(equal(set))
        expect(access.object(for: "newSet", ofClass: NSSet.self) as? Set).to(equal(newSet))
        expect(access.object(for: "existingDic", ofClass: NSDictionary.self) as? Dictionary<String, String>).to(equal(["key":"existingValue"]))
        expect(access.object(for: "newDic", ofClass: NSDictionary.self) as? Dictionary<String, String>).to(equal(["key":"newValue"]))
        expect(access.string(for: "existingString")).to(equal("existingValue"))
        expect(access.string(for:"newString")).to(equal("newValue"))
      }
      
      it("sets the value for new keys") {
        expect(access.register(defaults: ["newKey": NSNumber(value: true)])).to(beTrue())
        expect(access.bool(for: "newKey")).to(beTrue())
      }
      
      it("can set multiple keys") {
        let dic = ["foo": NSNumber(value: true), "bar": NSNumber(value: 100)]
        expect(access.register(defaults: dic)).to(beTrue())
        expect(access.bool(for: "foo")).to(beTrue())
        expect(access.integer(for: "bar")).to(equal(100))
      }
    }
    
    
  }
}

// MARK: - Private Helper Functions
private func data(for object: Any) -> Data? {
  if let data: Data = try? NSKeyedArchiver.archivedData(with: object) {
    return data
  }
  
  return nil
}


extension LUKeychainAccessSpec: KeychainServicesMockObserver {
  func addWasCalled(data: Data, key: String) {
    servicesObserver.record[Observer.Key.Add] =
    [Observer.Key.Data: data, Observer.Key.Key: key]
  }
  
  func dataWasCalled(key: String) {
    servicesObserver.record[Observer.Key.GetData] =
    [Observer.Key.Key: key]
  }
  
  func deleteAllItemsWasCalled() {
    servicesObserver.record[Observer.Key.DeleteAll] = [:]
  }
  
  func deleteItemWasCalled(key: String) {
    servicesObserver.record[Observer.Key.DeleteItem] =
    [Observer.Key.Key: key]
  }
  
  func updateWasCalled(data: Data, key: String) {
    servicesObserver.record[Observer.Key.Update] =
    [Observer.Key.Data: data, Observer.Key.Key: key]
  }
}
